### ISSUE TITLE
Would using a model different than VGG be a good idea?

### ISSUE THREAD

**Issue by mrakgr** [nkolkin13/NeuralNeighborStyleTransfer#10]
I have a 4gb GTX 970, but I want to generate 1k images using NNST so I am thinking how I could do that. My first idea would be to implement checkpointing, but I realized that VGG is narrow at the bottom, but wide at the top so rather than a 4x reduction in memory use, it would be more like 2x which is not enough to get me to 1k. I haven't tried your method yet, but I saw a Youtube vid which said it needed 12gb for 1024 x 1024 images. It also said that 6gb are required for 512 x 512 which is probably an error, more likely it needs only 3gb. Is that right?

Anyway, VGG is an old and big architecture, I am looking at more recent ones like EfficientNet and ConvNeXt which should have a better computational design and am considering adapting NNST to them, as well as potentially implementing checkpointing.

Before I undertake that kind of work, I just wanted to ask if you've tried out NNST with different models? If you have and then found that VGG is better, I'd like to hear about that. If not, then it might be worth trying NNST with a [ConvNeXt](https://pytorch.org/vision/stable/models.html#id59).

**Comment by mrakgr** [comment]
Actually, I might be wrong about VGG being narrow at the bottom and wide at the top. I see in the EfficientNets paper that they downsample the resolution while doubling the number of channels as the data goes up the stages. It makes zero sense that modern nets would be designed so the upper parts do most of the computation, so the same probably applies to the old and big VGG. In that case, checkpointing would be a good idea. I've looked at the ConvNext paper and saw that even the tiny version is bigger than VGG, so a lower memory burden is probably not a reason to pick it over VGG. EfficientNets are indeed smaller.

Edit: Also I've neglected to think about the intra block savings of checkpointing. This should be at least 3x. So doing checkpointing every 2 blocks should lead to a 6x reduction in memory consumption in a VGG net, this should be enough to get this method to run on a 4gb GPU.

**Comment by mrakgr** [comment]
The memory scaling is really confusing for this program. I've set the pyramidal levels down to 6 in order to get it to run with sizes smaller than 512, and modified the size option to test it. I've modified the code to print out the max memory allocated as well as reserved by the allocator.

```py
# Free gpu memory in case something else needs it later
if misc.USE_GPU:
    print(torch.cuda.max_memory_allocated() / 2 ** 30)
    print(torch.cuda.max_memory_reserved() / 2 ** 30)
    torch.cuda.empty_cache()
```

I'd expect the memory use would scale linearly with image size, but here is what I get with 128 x 128 images.

```
Done! total time: 25.85118007659912
0.7709341049194336
1.1171875
```

With 256 x 256.

```
Done! total time: 52.379212379455566
1.113663673400879
1.7421875
```

With 320 x 320. I expected it to increase by 56%, but it went up more than 2x.

```
Done! total time: 156.9229395389557
2.3323493003845215
2.7109375
```

I can't allocate more than 3gb on my GPU, so this size is my limit on this system. Given the wonky scaling, 6gb for 512 x 512 and 12gb 1024 x 1024 that I see in the docs seem less like an error now. I'll do more investigation into why this is so, I think that the NN itself should scale linearly with image size. I'll make sure of that next.

**Comment by nkolkin13** [comment]
Hi Marko,

I haven't tested other networks extensively as feature extractors. I did a few experiments with resnet (and found VGG to produce higher quality), but nothing with ConvNext or EfficientNet. While using a more memory efficient feature extractor would help, it's worth keeping in mind is that there are some additional memory bottlenecks here if you want to make a memory efficient version.

One is that part of the algorithm uses hyper-columns, where feature maps from multiple scales are upsampled and concatenated, this feature tensor itself can be a memory bottleneck (and depends only on resolution + the channel count of the selected layers, not the architecture). The main ways to reduce memory consumption here are using lower resolution feature maps, or feature maps with fewer channels.

In my experience the other major bottleneck is the distance matrix that's used to find matches between feature vectors, which is only dependent on the resolution of the feature maps. There are some tradeoffs you can make here though. One is to randomly subsample vectors from the style image (trading off quality for speed/memory). The other is finding matches for only a chunk of the content image at a time (trading off speed for memory, but not changing the final result).

Best,
Nick

**Comment by mrakgr** [comment]
I've spent a day walking through the code and my conclusion is that it would be very difficult to make it work with 4gb of memory. My initial assumption was that I could come in and add some checkpointing, this would be enough to deal with vanilla style transfer, but now that I've gone through the NNST codebase carefully, I see that there is a large amount of memory fragmentation due to all the intermediate operations happening. I couldn't get it to go through even just the forward pass due to that. I haven't realized it before, but PyTorch's allocator is kind of shitty, I managed to get through an obstacle just by clearing the cache which acts as manual defrag. PyTorch needs auto defrag capability instead of throwing an OOM error as soon as it runs out of space.

I thought I might find issues with code itself, but I could not and I do not see a way forward without algorithmic simplifications. In order to really improve on it I'd need to start with vanilla style transfer and check the NNST improvements one by one. Since this would be an engineering endeavor taking anywhere between a few weeks to months, I'll consider renting a beefy GPU on the cloud instead.

Regarding the question in the title, I am not sure it would be worth it to get a net with a smaller number of channels than VGG. If I cut the number channels by 4x there wouldn't be many of them left. CNNs don't have many parameters, it is the way they process images that makes it hard to pass large ones into them.

At this point it makes a lot more sense to just get better hardware.

**Comment by nkolkin13** [comment]
Also, this work was recently posted, and the results are quite impressive IMO. I haven't looked at their code or run it, so I don't know the memory requirements, but you may find it useful. https://github.com/zyxElsa/CAST_pytorch

**Comment by mrakgr** [comment]
Thank you, I'll check it out. Also, sorry for talking over you in the last post, I saw your post come in just as I was about to click submit, but was to lazy to redo it. If it was the old me, I wouldn't mind spending weeks and more playing around, but now I am into 3d art so that is what I'll focus on. Unless I put in a lot of effort into texturing, the 3d scenes come out rather plastic, so I'll be using style transfer to make them interesting.

One immediate market use that I can think of for your kind of work is turning photographs into [drawn backgrounds](https://www.youtube.com/watch?v=fNQVdfRfMWg). A lot of mangaka do it, and your method is impressively good at making drawings. It probably makes more sense to use NN style transfer than Photoshop/Clip Studio filters as the starting point for this kind of work. Clean, sharp drawings are versatile and it is easy to draw stylized characters into them.

**Comment by mrakgr** [comment]
I checked it out and it works pretty well. I posted a bunch of samples on [my Twitter](https://twitter.com/Ghostlike/status/1533813381442019329) if you are interested. Thank you for introducing me to it.

I had to make a few changes to the existing code, but I can get it to produce 1404 x 1404 images without going over the memory limit. As far as I can see the main strength of CAST is that it's good at preserving the original image content, but it has the weakness of not being able to go deeper in the stylized direction. I wouldn't say NNST is worse, it just depends on what you are going for, as well as the available computational resources, and the copyright status of the style images.
